Fix #4422: Add search option in library for mobile users

saching97 opened the pull request:
## Explanation
Fixes #4422   
Added a div with the search bar under the "Begin your adventure..." text, as discussed in the issue thread. This search bar will appear only on mobile devices and in all other cases remains hidden as the nav-bar already has a search bar on larger devices.

## Screenshots
**Mobile View Before**
![mobile-before](https://user-images.githubusercontent.com/24409931/37427076-3d2123f6-27ee-11e8-8f21-7243225d0d3b.png)

**Mobile View After**
![mobile-after](https://user-images.githubusercontent.com/24409931/37427090-42c9d03c-27ee-11e8-9c6d-2e7435db832e.png)

**Tablet View**
![tablet-view](https://user-images.githubusercontent.com/24409931/37427097-491e9d28-27ee-11e8-96c8-e9532d7ec750.png)

saching97 committed:
Fixes #4422

saching97 committed:
Fixed whitespacing

apb7 commented:
Hi @saching97! Thank you for your interest! Just to confirm, have you gone through [these](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#setting-things-up) instructions and filled the CLA as mentioned?

saching97 commented:
Hi @apb7 Yes, I've followed the instructions and signed the individual CLA

apb7 commented:
Great! Thanks, welcome to Oppia :)

seanlip commented:
Adding @christineycc @rachelwchen for comments on UI.

kevinlee12 reviewed (commented):
Hi @saching97, code-wise it is fine. Though on half screen, it looks like: 

![image](https://user-images.githubusercontent.com/5758256/37441033-99177e6c-27bc-11e8-8b8e-b4ec436d2f4d.png)

Perhaps a behavior that youtube does might be better (the search bar takes up the entire navbar).

saching97 commented:
Hi @kevinlee12 ,
You're right, I missed that - changing `<div class="visible-xs" style="margin: 0 auto; width: 75%;">` to `<div class="visible-xs" style="margin: 0 auto; width: fit-content;">` centrally aligns it properly.

![modified-mobile](https://user-images.githubusercontent.com/24409931/37449029-c8b856ae-284e-11e8-87f1-37663daedc08.png)

Should we proceed in this way or go for a persistent search bar, similar to YouTube's approach?

seanlip commented:
Those are cool science lessons :D 

I'd suggest waiting for design input from @rachelwchen or @christineycc, they can suggest what's best. Thanks!

rachelwchen commented:
Thanks for filing this issue @saching97! Welcome to Oppia :) 

Three comments:
1. Placement - 
Having the search bar there is good. Another alternative would be to put a search icon on the nav bar, so that it doesn't take up vertical screen estate. I'm not sure which option is better, though! Just throwing this idea out there.

2. Color - 
I think I'd prefer the background color to be white. Having a dark green rectangle in the middle against a blue background seems a little off.

3. Size -
I think the interface would look cleaner if the width of the search bar aligns with the big text. Screenshot attached:
![mobile portrait](https://user-images.githubusercontent.com/12159451/37490637-8ca8f22c-2871-11e8-9c8b-16112ecfbdda.png)

saching97 commented:
Thanks for the review @rachelwchen . I'm increasing the search bar width from 300px to 375px. This aligns the search bar to the upper heading's width in mobile devices. In tablet view/half screen view the heading is too large and having such a wide search bar seemed odd to me, so I have kept the width fixed to 375px in that case as well.

Here I'm attaching the screenshots of changing the colour of the search bar to faded white and transparent. I think that the faded white looks good but I'll wait for the comments before committing the change.

As for the placement, we can change it if needed. I was following the discussion on the main issue and thought this was easy enough to implement and went forward with it. One drawback of this layout I feel is that modifying searches will require the person to go back to the library main page and then search again. 

**Faded White**
![faded white background](https://user-images.githubusercontent.com/24409931/37518034-826fc254-2939-11e8-8662-2b7f468e30d2.png)

**Transparent**
![transparent background](https://user-images.githubusercontent.com/24409931/37518038-84653a12-2939-11e8-9879-a7b7157802f2.png)

**Tab View**
![tabview](https://user-images.githubusercontent.com/24409931/37518042-862faf9e-2939-11e8-8e33-1dca9dea4dc1.png)

saching97 committed:
Fixed centre alignment

rachelwchen commented:
Looking good! I agree with you - faded white looks better. Though to follow Oppia's existing design, using full colors (ie not faded or transparent) might be more consistent. Thank you!

saching97 committed:
Changed size and background color

saching97 commented:
@rachelwchen 
I've committed the width change and changed the background colour to white. Here are the screenshots:

**White Background (Small screen)**
![white background](https://user-images.githubusercontent.com/24409931/37594996-1a5c753e-2b9d-11e8-8b7b-08717e2c579c.png)

**White Background (Tablet/Half-Screen)**
![tab view white](https://user-images.githubusercontent.com/24409931/37595109-869922ce-2b9d-11e8-8b51-ece0c03575b9.png)

**Desktop View**
![unchanged desktop view](https://user-images.githubusercontent.com/24409931/37595032-4015d590-2b9d-11e8-9ffd-c97fb29f77ea.png)

christineycc commented:
@saching97 I think these looks great! One small thing, I think it would be cool to add the "all categories filter" to the search bar on the tablet version, and align it with "begin your adventure by browsing our". Otherwise, these looks good.

saching97 committed:
Included Category bar for tablets

saching97 commented:
@christineycc I've added the category bar for tablets (screenshot below). I didn't align it with the "begin your.." exactly because it was causing several alignment issues with the search bar on different screen sizes. If necessary, I can try it again, but the change seemed fine imo:

![include categories](https://user-images.githubusercontent.com/24409931/37655228-b312bd74-2c6a-11e8-8917-24c2d900b988.png)

saching97 committed:
Fixed dropdown spacing

seanlip commented:
/cc @christineycc -- ping?

christineycc commented:
Sorry for the late reply! But yes, The change looks good to me

kevinlee12 commented on core/templates/dev/head/pages/library/search_bar_directive.html:
```diff
@@ -131,8 +134,50 @@
+      background: #ffffff;
+      color: #04857c;
+    }
+
```
Remove empty line.

kevinlee12 commented on core/templates/dev/head/pages/library/search_bar_directive.html:
```diff
@@ -131,8 +134,50 @@
+    .oppia-search-bar-input:-ms-input-placeholder {
+      color: #04857c;
+    }
+
```
Ditto

kevinlee12 commented on core/templates/dev/head/pages/library/search_bar_directive.html:
```diff
@@ -131,8 +134,50 @@
+      background: #ffffff;
+      color: #04857c;
+    }
+
```
Ditto

kevinlee12 commented on core/templates/dev/head/pages/library/search_bar_directive.html:
```diff
@@ -131,8 +134,50 @@
+      margin-top: -64px;
+      margin-left: 10px;
+    }
+
```
Ditto

kevinlee12 commented on core/templates/dev/head/pages/library/search_bar_directive.html:
```diff
@@ -131,8 +134,50 @@
+    .oppia-search-bar-input.btn:hover, .oppia-search-bar-input.btn:focus {
+      color: #04857c;
+    }
+
```
Ditto

kevinlee12 commented on core/templates/dev/head/pages/library/search_bar_directive.html:
```diff
@@ -131,8 +134,50 @@
+    .protractor-test-search-bar-dropdown-menu{
+      margin-top:-25px;
+    }
+
```
Ditto

kevinlee12 commented on core/templates/dev/head/pages/library/search_bar_directive.html:
```diff
@@ -131,8 +134,50 @@
+      margin-top:-25px;
+    }
+
+    .oppia-same-row{
```
Space after `row`

kevinlee12 commented on core/templates/dev/head/pages/library/search_bar_directive.html:
```diff
@@ -131,8 +134,50 @@
+      color: #04857c;
+    }
+
+    .protractor-test-search-bar-dropdown-menu{
```
Ditto

seanlip commented on core/templates/dev/head/pages/library/search_bar_directive.html:
```diff
@@ -131,8 +134,50 @@
+      color: #04857c;
+    }
+
+    .oppia-search-bar-input.btn{
```
Here and below, add space before {

kevinlee12 replied:
Ditto

seanlip commented on core/templates/dev/head/pages/library/search_bar_directive.html:
```diff
@@ -131,8 +134,50 @@
+    }
+  }
+  @media (max-width: 629px) {
+  .oppia-search-bar-text-input {
```
Should be indented by 2 from the previous line

kevinlee12 commented on core/templates/dev/head/pages/library/search_bar_directive.html:
```diff
@@ -131,8 +134,50 @@
+       display: inline-block;
+    }
+  }
+  @media (max-width: 629px) {
```
Add empty line before this.

kevinlee12 commented on core/templates/dev/head/pages/library/search_bar_directive.html:
```diff
@@ -131,8 +134,50 @@
+    }
+
+    .protractor-test-search-bar-dropdown-menu{
+      margin-top:-25px;
```
Space after `:`

seanlip commented on core/templates/dev/head/pages/library/search_bar_directive.html:
```diff
@@ -1,49 +1,52 @@
-        </span>
-      </div>
-      <input type="text" class="form-control oppia-search-bar-input oppia-search-bar-text-input protractor-test-search-input"
+<div class="oppia-same-row">
```
I don't really understand the name of this class. Was it not in the same row before?

seanlip reviewed (commented):
Thanks @christineycc!

@saching97 -- I took a code pass. Could you PTAL and also fix the merge conflicts?

kevinlee12 commented on core/templates/dev/head/pages/library/search_bar_directive.html:
```diff
@@ -131,8 +134,50 @@
+    .oppia-search-bar-input.btn{
+      color: #04857c;
+      margin-top: -64px;
+      margin-left: 10px;
```
Put this above `margin-top`, we need to keep this alphabetized,

kevinlee12 commented on core/templates/dev/head/pages/library/search_bar_directive.html:
```diff
@@ -1,49 +1,52 @@
 
-<div uib-dropdown class="navbar-left oppia-navbar-button-container oppia-search-bar-category-selector protractor-test-search-bar-category-selector">
-  <button uib-dropdown-toggle type="button" class="btn protractor-test-search-bar-dropdown-toggle oppia-search-bar-input ng-cloak"
+<div class="oppia-same-row">
```
Please rename as this isn't very descriptive that it's a container.

kevinlee12 reviewed (changes requested):
Hi @saching97, left some comments, ptal!

kevinlee12 commented:
^ Please also resolve the merge conflicts!

saching97 committed:
Fixed spacing errors

saching97 committed:
Merge branch 'develop' into library-mobile-search

saching97 commented:
Thanks for the reviews! I've made all the changes as requested and resolved the merge conflict, PTAL. The `oppia-same-row-container` class is necessary to place the category and search bar in the same row inside a div since:
`<div class="visible-xs" style="margin: 0 auto; width: fit-content;"><search-bar></search-bar></div>` pushes the category bar to the next row.
In the nav-bar implementation, this is not required since the search bar is not present inside a div.

seanlip commented:
I just tested this. After you search for something on a small screen device, the search results show, but the search bar disappears. Is that intentional?
![screenshot from 2018-03-26 16-25-17](https://user-images.githubusercontent.com/10575562/37938378-634c5a50-3112-11e8-9fda-b98c0e60693e.png)

saching97 committed:
Added search bar to results page

saching97 committed:
Added search bar to results page

saching97 committed:
Fixed change of search bar in results

kevinlee12 commented on core/templates/dev/head/pages/library/search_bar_directive.html:
```diff
@@ -131,12 +134,47 @@
+       display: inline-block;
+    }
+  }
+  @media (max-width: 629px) {
```
Add newline above

kevinlee12 reviewed (dismissed):
Hi @saching97, small code nit. As for the UI, there appears to be some issues (though the problem that Sean reported is no longer showing):

Tablet:
The categories tab drops to the bottom:
![image](https://user-images.githubusercontent.com/5758256/38058824-0bc3b32c-3299-11e8-96aa-0030209cd3f6.png)

Phones (iPhone SE width):
The search bar bleeds off the screen.
![image](https://user-images.githubusercontent.com/5758256/38058676-8ba20b08-3298-11e8-8291-8a12c647eb98.png)

saching97 committed:
Fixed small-screen overflow and tablet-view alignment

saching97 committed:
Merge remote-tracking branch 'origin/library-mobile-search' into library-mobile-search

saching97 commented:
Thanks for pointing that out @seanlip , I've added the search bar in the results page as well (screenshot below).
Hi @kevinlee12 , I've fixed the UI issues (screenshots below) and made the code modifications. For my reference, is there any documentation I can follow to prevent such linting errors? [Coding style guidelines](https://github.com/oppia/oppia/wiki/Coding-style-guide) lack some of these intricacies and the lint test doesn't show any issue in the pre-submit-check.

**Results Page**
![results](https://user-images.githubusercontent.com/24409931/38111853-ce798d0e-33bd-11e8-92dc-97f8e7526b19.png)

**Small screen devices (iPhone 4)**
![small screen](https://user-images.githubusercontent.com/24409931/38111855-cedb3ae0-33bd-11e8-93e3-90fda753b0db.png)

**Tablet mode (iPad)**
![tablet mode](https://user-images.githubusercontent.com/24409931/38111856-cf388ca4-33bd-11e8-8463-4038301bb38c.png)

apb7 commented:
Hi @saching97! 
Regarding lint checks, we are in the process of implementing more pre-submit checks for HTML files so hopefully, you won't face this problem in the near future. For reference, you can go through the existing HTML files in the codebase. 
Thanks :)

kevinlee12 commented on core/templates/dev/head/pages/library/library.html:
```diff
@@ -85,6 +85,10 @@ <h1 class="oppia-library-h1" translate="I18N_LIBRARY_MAIN_HEADER"></h1>
             <h2 class="oppia-library-h2" translate="I18N_LIBRARY_SUB_HEADER"></h2>
           </div>
 
+          <div class="visible-xs" style="margin: 0 auto; width: fit-content; padding-right:15px">
```
Add a space after the last `:`.

kevinlee12 reviewed (commented):
Hi @saching97, one more UI nit:

At 682px on desktop:
Search bar should be longer. 
![image](https://user-images.githubusercontent.com/5758256/38114693-37fa60ec-335e-11e8-81e9-c97fe92e4df0.png)

kevinlee12 commented:
I'll be out this weekend so I'll hand over the review to @seanlip.

saching97 commented:
Hi @kevinlee12 @seanlip  - I'm unable to reproduce mentioned screenshot, can you please elaborate on what dimensions you used? I'm getting the following:

**682x682**
![cannot repro](https://user-images.githubusercontent.com/24409931/38132235-dc265940-3427-11e8-8f53-f53cf0052c7b.png)

**386x682**
![cannot repro 2](https://user-images.githubusercontent.com/24409931/38132236-dc7f8786-3427-11e8-99f8-c54d4a4d5aec.png)

seanlip commented:
@kevinlee12's out for the weekend -- @nithusha21, would you mind taking a look please?

nithusha21 commented:
I can't repro the screenshot posted by @kevinlee12 either. One question regarding functionality though. So the category/language selection is hidden for smaller screen sizes. The default language in the search bar is English. Shouldn't the user be allowed to select this? 

For example, you are viewing the library page in German. On searching, the language_code=("en") parameter is still passed in the URL. The user interface gives the user no way to select the language of the exploration he is looking for (Which means that he will never be able to find a german exploration on mobile). Not sure if hiding the language dropdown is a viable solution, or at least set it to all languages so that the user is at least viewing the search results of languages other than English. 

Any views on this @seanlip, @saching97, @rachelwchen

saching97 committed:
Added missing space in code

saching97 commented:
Hi @nithusha21 The language drop-down gets hidden for all screen-sizes with width lesser than 1024px in the current implementation. While the default search language is English, if the user wishes to switch to any other language, they can do so through Preferences > Preferred Exploration Languages.

![choose language](https://user-images.githubusercontent.com/24409931/38137270-6b20cc52-3441-11e8-82e6-f903ee6ca518.png)

seanlip commented:
@nithusha21 -- good point about language. I've thought about it and I think we'll leave that for a later date, since the library page is due for a more general redesign anyway and I don't think it needs to block this PR (especially since the language selector is already being hidden anyway for small screen sizes).

Since @kevinlee12's bug seems tricky to repro I'll go ahead and merge this. @kevinlee12 -- when you come back, perhaps you could open a new issue with more details on how to repro your screenshot?

Also, thanks @saching97 for your PR!
